apps/prairielearn/src/pages/instructorFileEditor/instructorFileEditor.js

```diff
@@ -466,616 +453,148 @@ async function writeEdit(fileEdit) {
-    };
+      const fullPath = path.join(fileEdit.coursePath, fileEdit.dirName, fileEdit.fileName);
+      const contents = await fs.readFile(fullPath, 'utf8');
+      fileEdit.diskHash = getHash(b64Util.b64EncodeUnicode(contents));
+      if (fileEdit.origHash !== fileEdit.diskHash) {
+        throw new Error(`Another user made changes to the file you were editing.`);
+      }
 
-    const _didSave = () => {
-      debug(`${job_sequence_id}: _didSave`);
-      const jobOptions = {
-        course_id: options.course_id,
-        user_id: options.user_id,
-        authn_user_id: options.authn_user_id,
-        type: 'did_save',
-        description: 'Mark edit as saved',
-        job_sequence_id: job_sequence_id,
-        on_success: _unlock,
-        on_error: _cleanup,
-        no_job_sequence_update: true,
-      };
-      serverJobs.createJob(jobOptions, (err, job) => {
-        if (ERR(err, (err) => logger.info(err))) {
-          _finishWithFailure();
-          return;
-        }
+      await fs.writeFile(fullPath, b64Util.b64DecodeUnicode(fileEdit.editContents), 'utf8');
+      job.verbose(`Wrote changed to ${fullPath}`);
 
-        debug('Mark edit as saved');
-        updateDidSave(fileEdit, (err) => {
-          if (err) {
-            job.fail(err);
-          } else {
-            job.verbose('Marked edit as saved');
-            job.succeed();
-          }
-        });
-      });
-    };
-
-    const _unlock = () => {
-      debug(`${job_sequence_id}: _unlock`);
-      const jobOptions = {
-        course_id: options.course_id,
-        user_id: options.user_id,
-        authn_user_id: options.authn_user_id,
-        type: 'unlock',
-        description: 'Unlock',
-        job_sequence_id: job_sequence_id,
-        on_success: jobSequenceHasFailed ? _finishWithFailure : _getEndCommitHash,
-        on_error: _finishWithFailure,
-        no_job_sequence_update: true,
-      };
-      serverJobs.createJob(jobOptions, (err, job) => {
-        if (ERR(err, (err) => logger.info(err))) {
-          _finishWithFailure();
-          return;
+      if (config.fileEditorUseGit) {
+        try {
+          await job.exec('git', ['reset'], {
+            cwd: locals.course.path,
+            env: gitEnv,
+          });
+          await job.exec('git', ['add', path.join(fileEdit.dirName, fileEdit.fileName)], {
+            cwd: locals.course.path,
+            env: gitEnv,
+          });
+          await job.exec(
+            'git',
+            [
+              '-c',
+              `user.name="${fileEdit.user_name}"`,
+              '-c',
+              `user.email="${fileEdit.uid}"`,
+              'commit',
+              '-m',
+              fileEdit.commitMessage,
+            ],
+            {
+              cwd: locals.course.path,
+              env: gitEnv,
+            },
+          );
+        } catch (err) {
+          await job.exec('git', ['checkout', path.join(fileEdit.dirName, fileEdit.fileName)], {
```

Comment:
`git reset --hard` might be more appropriate?

Reply:
See https://github.com/PrairieLearn/PrairieLearn/pull/7966#discussion_r1261719309 and #8177.